### PR TITLE
Thread pool

### PR DESCRIPTION
**Pull request by Dekz120**

### CODE REVIEW

File: elooplib/archive.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#pragma once
+#  define SET_BINARY_MODE(file)
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include "zlib.h"
```
Comment (Yanpas) [inline elooplib/archive.hpp:6]: Use angle brackets for system libraries

File: Makefile [hunk 2]
```diff
@@ -1,31 +1,34 @@
...
-all: bvnserver clean
+#all: bvnserver clean
 
 bvnserver: main.o eventloop.a
-	$(CXX) $(CXXFLAGS) main.o eventloop.a -o bvnserver
+	$(CXX) $(CXXFLAGS) main.o eventloop.a -o $@
 
-main.o: $(SOURCES)
-	$(CXX)  -c $(CXXFLAGS) -I./elooplib $(SOURCES) 
+main.o: $(SOURCES) $(HEADERS)
```
Comment (Yanpas) [inline Makefile]: Please use %.o patterns, DRY - do not repeat yourself

File: Makefile [hunk 3]
```diff
@@ -1,31 +1,34 @@
...
+HEADDIR = ./elooplib
+HEADERS = ./elooplib/*.hpp
+OBJFILES = eventloop.o client.o server.o node.o
 
 CXX = g++
-CXXFLAGS = -std=c++17 -g -Wall -Wextra #-Werror
+CXXFLAGS = -std=c++17 -g -fsanitize=address -Wall -Wextra #-Werror
 
-all: bvnserver clean
+#all: bvnserver clean
```
Comment (Yanpas) [inline Makefile]: Why did you remove all? Btw all must not include clean

File: Makefile [hunk 4]
```diff
@@ -1,31 +1,34 @@
-LSOURCES= elooplib/node.cpp elooplib/server.cpp elooplib/client.cpp elooplib/eventloop.cpp
-SOURCES = bvnServer/main.cpp
-HEADERS = ./elooplib
+LSOURCES= elooplib/server.cpp elooplib/client.cpp elooplib/eventloop.cpp
+SOURCES = elooplib/threadpool.hpp bvnServer/main.cpp
+HEADDIR = ./elooplib
+HEADERS = ./elooplib/*.hpp
```
Comment (Yanpas) [inline Makefile]: HEADERS defined twice

File: elooplib/archive.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+static constexpr size_t chunk = 16384;
+int compress(FILE *source, FILE *dest, int level)
```
Comment (Yanpas) [inline elooplib/archive.hpp]: Functions inside headers should be either inline or static. But that's not the case, please use cpp file

File: elooplib/archive.hpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    strm.zalloc = Z_NULL;
+    strm.zfree = Z_NULL;
+    strm.opaque = Z_NULL;
+    ret = deflateInit(&strm, level);
+    if (ret != Z_OK)
+        return ret;
+
+    do {
+        strm.avail_in = fread(in, 1, CHUNK, source);
+        if (ferror(source)) {
```
Comment (Yanpas) [inline elooplib/archive.hpp]: Use return codes instead of ferror

File: elooplib/archive.hpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            (void)deflateEnd(&strm);
```
Comment (Yanpas) [inline elooplib/archive.hpp]: Forgot to check return code?

File: elooplib/archive.hpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                (void)deflateEnd(&strm);
+                return Z_ERRNO;
+            }
+        } while (strm.avail_out == 0);
+        assert(strm.avail_in == 0);     
+
+    } while (flush != Z_FINISH);
+    assert(ret == Z_STREAM_END);       
+
+    (void)deflateEnd(&strm);
```
Comment (Yanpas) [inline elooplib/archive.hpp]: Have you written this code yourself ? 😏

File: elooplib/client.cpp [hunk 9]
```diff
@@ -35,32 +37,95 @@ int Client::handleConnection()
...
+        return;
+    }
+
+    int dir_len = 0;
+    // TODO try_lock if(1) etc
+    std::list<std::string> files;
+    for (const auto &entry : std::filesystem::directory_iterator(response))
+    {
+        std::string tmp = entry.path();
+        if(tmp.find(".gz") == std::string::npos)
```
Comment (Yanpas) [inline elooplib/client.cpp]: C++20 has ends_with

File: elooplib/client.cpp [hunk 10]
```diff
@@ -35,32 +37,95 @@ int Client::handleConnection()
...
+    const std::filesystem::path dir(response);
+    if (!exists(dir))
+    {
+        response = "Path is not exists\n";
+        return;
+    }
+
+    int dir_len = 0;
+    // TODO try_lock if(1) etc
+    std::list<std::string> files;
```
Comment (Yanpas) [inline elooplib/client.cpp]: Why list instead of vector? Why do you need container at all?

File: elooplib/client.cpp [hunk 11]
```diff
@@ -35,32 +37,95 @@ int Client::handleConnection()
...
+        }
+
+        dir_len++;
+    }
+    
+    files_counter = 0;
+    for(const auto &f: files)
+        compressFile(f);
+
+    response = std::to_string(files_counter);
```
Comment (Yanpas) [inline elooplib/client.cpp]: Currently whole event loop is blocked until every file is not compressed, you should redesign your event loop to avoid it. Take a look at std future

File: elooplib/eventloop.cpp [hunk 12]
```diff
@@ -93,6 +93,7 @@ void EventLoop::run()
         }
         std::signal(SIGINT, EventLoop::handleSignal);
```
Comment (Yanpas) [inline elooplib/eventloop.cpp]: You should call signal only once!

File: elooplib/threadpool.hpp [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+    ThreadPool(ThreadPool&) = delete; 
+    ThreadPool& operator=(ThreadPool&) = delete; 
+    ThreadPool(ThreadPool&&) = delete;
+    ThreadPool& operator=(ThreadPool&&) = delete; 
+                                   //threads(std::move(rhs.threads)){};
+
+    template <typename Func, typename... Args>
+    void add_task(const Func &task_func, Args &&...args)
```
Comment (Yanpas) [inline elooplib/threadpool.hpp]: Wow 😳

**Review by Yanpas — CHANGES_REQUESTED** [review]

File: elooplib/archive.hpp (hunk 8 above)
Comment (Dekz120) [reply elooplib/archive.hpp]: No
NO

**Comment by d84833** [comment]
It's so deep!

**Comment by Dekz120** [comment]
> It's so deep!

Yeah, i see that.

File: Makefile [hunk 14]
```diff
@@ -19,16 +17,20 @@ eventloop.a: $(OBJFILES)
...
 	
 eventloop.o: $(LSOURCES) 
-	$(CXX) -c $(CXXFLAGS)  elooplib/eventloop.cpp
+	$(CXX) -c $(CXXFLAGS) elooplib/eventloop.cpp
 	
-client.o: elooplib/client.cpp elooplib/node.cpp elooplib/threadpool.hpp elooplib/archive.hpp
+client.o: elooplib/client.cpp elooplib/node.cpp elooplib/threadpool.cpp elooplib/archive.hpp
 	$(CXX) -c $(CXXFLAGS) elooplib/client.cpp
 	
 server.o: elooplib/server.cpp elooplib/node.cpp
```
Comment (Yanpas) [inline Makefile]: No need to recompile server.cpp if node.cpp changes. These files are independent. Make it depend on `node.hpp` instead.

Remember, first you compile every cpp file to o file, then link all o files to single executable. Header files don't get compiled, they are only included to cpp files and usually have only declaration.

File: elooplib/client.cpp [hunk 15]
```diff
@@ -1,131 +1,187 @@
...
-        Client::handleTime();
-        return;
-    }
-
-    if (command == "echo ")
-    {
-        Client::handleEcho();
-        return;
-    }
+  if (response.starts_with("time\n"))
```
Comment (Yanpas) [inline elooplib/client.cpp:42]: Let's introduce separate member called `request`. To improve readability.

File: elooplib/client.cpp [hunk 16]
```diff
@@ -1,131 +1,187 @@
...
+{
+  response = response.substr(9, request_field.size() - 10);
+  const std::filesystem::path dir(response);
+  if (!exists(dir))
+  {
+    response = "Path is not exists\n";
+    return 0;
+  }
+  int fd = eventfd(0, EFD_CLOEXEC);
+  return fd;
```
Comment (Yanpas) [inline elooplib/client.cpp]: Why do you return fd? return value of all functions is treated as if thread pool needs to be used. Return 1 instead. Or better use bool as a return value.

File: elooplib/client.cpp [hunk 17]
```diff
@@ -1,131 +1,187 @@
...
+int Client::handleCompress()
+{
+  response = response.substr(9, request_field.size() - 10);
+  const std::filesystem::path dir(response);
+  if (!exists(dir))
+  {
+    response = "Path is not exists\n";
+    return 0;
+  }
+  int fd = eventfd(0, EFD_CLOEXEC);
```
Comment (Yanpas) [inline elooplib/client.cpp:90]: On error, -1 is returned and errno is set to indicate the error.

File: elooplib/client.cpp [hunk 18]
```diff
@@ -1,131 +1,187 @@
...
+  {
+    int s = send(getFd(), "Start compression...\n", 21, 0);
+    if (s < 0)
+      throw serverExcept("send()");
+    return r;
+  }
 }
 
-void Client::handleCompress()
+size_t Client::getFd() { Node::getFd(); };
```
Comment (Yanpas) [inline elooplib/client.cpp]: Your function returns undefined value. Please inspect warnings from the compiler. And let's make them treated as errors

File: elooplib/client.cpp [hunk 19]
```diff
@@ -1,131 +1,187 @@
...
+
+std::string Client::getResponse() { return response; };
```
Comment (Yanpas) [inline elooplib/client.cpp]: 1. Extra `;`
2. Make this method `const`
3. I guess you want to return by reference:
`const std::string& Client::getResponse() const { return response; }`

File: elooplib/client.cpp [hunk 20]
```diff
@@ -1,131 +1,187 @@
...
+//////////
+ClientTask::ClientTask(int fd, Client &cl, std::shared_ptr<ThreadPool> &tp)
```
Comment (Yanpas) [inline elooplib/client.cpp]: pass shared_pointer by copy, this class is intended to be used like that!

File: elooplib/client.cpp [hunk 21]
```diff
@@ -1,131 +1,187 @@
...
+  Node::closeConnection();
+  setFd(-1); // needed to keep socket opened after destructor call
 }
 
+size_t ClientTask::getFd() { return event_fd; }
+ClientTask::~ClientTask() { close(event_fd); }
+/////////
+CompressITask::CompressITask(const std::string &f, int fd, std::atomic<int> *tn,
+                             std::atomic<int> *s, int m)
+    : ITask(), filename(f), event_fd(fd), task_num(tn), success(s), max(m){};
```
Comment (Yanpas) [inline elooplib/client.cpp]: rename task_num, success and max to something more meaningful

File: elooplib/client.cpp [hunk 22]
```diff
@@ -1,131 +1,187 @@
...
 
-        }
+int ClientTask::compressFiles()
+{
 
-        dir_len++;
+  // TODO try_lock for directory name mutex
+  int task_summ = 0;
+  for (const auto &entry : std::filesystem::directory_iterator(dir))
+    task_summ++;
```
Comment (Yanpas) [inline elooplib/client.cpp]: why don't you check for `gz` ending here?

File: elooplib/client.hpp [hunk 23]
```diff
@@ -5,24 +5,56 @@
...
+    ~ClientTask();
+    private:
+    std::shared_ptr<ThreadPool> th_pool;
+    std::string dir;
+    int event_fd;
+    std::atomic<int> complete_tasks{0};
+    std::atomic<int> success_tasks{0};
 };
+
+class CompressITask : public ITask
```
Comment (Yanpas) [inline elooplib/client.hpp]: ZlibCompressTask. This class is not an interface anymore, so doesn't require `I` in it's name

File: elooplib/client.cpp [hunk 24]
```diff
@@ -1,131 +1,187 @@
...
+                             std::atomic<int> *s, int m)
+    : ITask(), filename(f), event_fd(fd), task_num(tn), success(s), max(m){};
+
+void CompressITask::run() { compressFile(); }
 
+int CompressITask::compressFile()
+{
+  // TODO compression
+  using namespace std::chrono_literals;
+  std::this_thread::sleep_for(6000ms); // DEBUG
```
Comment (Yanpas) [inline elooplib/client.cpp]: Tip: you can use defines for this purpose:
`#ifdef BVN_DEBUG`. And set it in Makefile optionally

File: elooplib/client.cpp [hunk 25]
```diff
@@ -1,131 +1,187 @@
...
 }
 
-void Client::compressFile(const std::string &in_file) // TODO
+int ClientTask::handleConnection()
 {
-    /*int res = compress(in_file.c_str(), (in_file + ".gz").c_str(),
-                       6);
-    if (res == Z_OK)
-        files_counter++; */
+  dir = std::to_string(success_tasks) + "\n";
```
Comment (Yanpas) [inline elooplib/client.cpp]: `auto response_body = `

File: elooplib/eventloop.cpp [hunk 26]
```diff
@@ -63,37 +62,76 @@ int EventLoop::handleConnection(std::shared_ptr<Node> &hst)
     return -1;
 }
 
+EventLoop::nodePointerType EventLoop::waitersSearch(int fd)
+{
+    auto nodeComp = [&fd](nodeType &n)
+    {
+        return fd == n.second->getFd();
+    };
+    auto wp = find_if(begin(wait_q), end(wait_q), nodeComp);
```
Comment (Yanpas) [inline elooplib/eventloop.cpp]: remove `using namespace std`

File: elooplib/eventloop.cpp [hunk 27]
```diff
@@ -63,37 +62,76 @@ int EventLoop::handleConnection(std::shared_ptr<Node> &hst)
...
+    return wp;
```
Comment (Yanpas) [inline elooplib/eventloop.cpp]: find_if returns iterator, not pointer. But vector's iterator gets automatically casted to pointer

File: elooplib/eventloop.cpp [hunk 28]
```diff
@@ -63,37 +62,76 @@ int EventLoop::handleConnection(std::shared_ptr<Node> &hst)
...
             throw serverExcept("poll()");
 
-        for (nodeRetType node_p = nodes.begin(); node_p != nodes.end(); node_p++)
+        for (nodePointerType node_p = nodes.begin(); node_p != nodes.end(); node_p++)
         {
-            auto &[pos, node] = *node_p;
+            auto [pos, node] = *node_p;
             if (fds[pos].revents == POLLIN)
             {
                 int data = handleConnection(node);
```
Comment (Yanpas) [inline elooplib/eventloop.cpp]: It's not data, it's `needsThreadPool`

Comment (Yanpas) [reply elooplib/eventloop.cpp]: Please make every return type meaningful. Introuduce type aliases. E.g.: 
`using fd_t = int; using error_t = int`

Comment (Yanpas) [reply elooplib/eventloop.cpp]: Also rename variables to appropriate names, code is unreadable

File: elooplib/eventloop.cpp (hunk 12 above)
Comment (Yanpas) [reply elooplib/eventloop.cpp]: Whe will you solve this?

File: elooplib/eventloop.hpp [hunk 29]
```diff
@@ -20,14 +20,20 @@ class EventLoop
...
+    std::list<std::pair<size_t, std::shared_ptr<Node>>> nodes; // TODO no pair, just remove from vector
+    std::list<std::pair<size_t, std::shared_ptr<Node>>> wait_q;
+    std::shared_ptr<ThreadPool> th_pool;
+
 public:
     EventLoop(size_t);
     void addNode(std::shared_ptr<Node> &, bool);
-    using nodeRetType = decltype(std::begin(nodes));
-    nodeRetType rmNode(nodeRetType &);
+    using nodeType = decltype(*std::begin(nodes));
```
Comment (Yanpas) [inline elooplib/eventloop.hpp:29]: use `decltype(nodes)::value_type` and so on:
https://en.cppreference.com/w/cpp/container/vector

File: elooplib/server.cpp [hunk 30]
```diff
@@ -30,3 +30,8 @@ int Server::handleConnection()
         throw serverExcept("accept()");
     return (accept_sd);
 }
+
+size_t Server::getFd()
+{
+    Node::getFd();
```
Comment (Yanpas) [inline elooplib/server.cpp]: Return value is undefined

File: elooplib/node.hpp [hunk 31]
```diff
@@ -38,7 +38,7 @@ class Node
     Node(Node &&);
     bool is_active();
     void closeConnection();
-    size_t getFd();
+    virtual size_t getFd();
```
Comment (Yanpas) [inline elooplib/node.hpp:42]: Make it pure virtual

File: elooplib/threadpool.cpp [hunk 32]
```diff
@@ -0,0 +1,46 @@
...
+      lock.unlock();
+
+      elem->run();
+    }
+  }
+}
+
+ThreadPool::~ThreadPool()
+{
+  for (uint32_t i = 0; i < threads.size(); ++i)
```
Comment (Yanpas) [inline elooplib/threadpool.cpp:40]: Why not `for(auto& t: threads)`?

Comment (Yanpas) [reply elooplib/threadpool.cpp:40]: Also `size_t` should be used for i

File: elooplib/threadpool.cpp [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+void ThreadPool::addTask(std::shared_ptr<ITask> &task)
+{
+  std::lock_guard<std::mutex> q_lock(q_mtx);
+  t_queue.push(task);
+  q_cv.notify_one();
+}
+
+void ThreadPool::run()
```
Comment (Yanpas) [inline elooplib/threadpool.cpp:19]: That's a canonical thread pool, very common pattern. Pay attention to understand how it works.

File: elooplib/threadpool.cpp [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+{
+  while (!quite)
```
Comment (Yanpas) [inline elooplib/threadpool.cpp:21]: `quit`

File: elooplib/threadpool.hpp [hunk 35]
```diff
@@ -4,86 +4,42 @@
...
-        {
-            threads.emplace_back(&ThreadPool::run, this);
-        }
-    }
-    ThreadPool(ThreadPool&) = delete; 
-    ThreadPool& operator=(ThreadPool&) = delete; 
-    ThreadPool(ThreadPool&&) = delete;
-    ThreadPool& operator=(ThreadPool&&) = delete; 
-                                   //threads(std::move(rhs.threads)){};
+    virtual void run(){};
```
Comment (Yanpas) [inline elooplib/threadpool.hpp]: Interface must have pure virtual methods:
`virtual void run() = 0;`

File: elooplib/threadpool.hpp [hunk 36]
```diff
@@ -4,86 +4,42 @@
...
 
-    template <typename Func, typename... Args>
-    void add_task(const Func &task_func, Args &&...args)
-    {
-        std::cout << "Add obj to threadpool\n";
-        std::lock_guard<std::mutex> q_lock(q_mtx);
-        t_queue.emplace(std::async(std::launch::deferred, task_func, args...));
-        q_cv.notify_one();
-    }
+    virtual ~ITask(){};
```
Comment (Yanpas) [inline elooplib/threadpool.hpp]: `virtual ~ITask() = default`

or
`virtual ~ITask();`
and in cpp file:
`ITask::~ITask() = default`

File: elooplib/threadpool.hpp [hunk 37]
```diff
@@ -4,86 +4,42 @@
...
-        if (!p_instance)
-        {
-            p_instance = std::shared_ptr<ThreadPool>(new ThreadPool(3));
-        }
-        return p_instance;
-    }
+class ThreadPool
+{
+public:
+    ThreadPool(size_t);
```
Comment (Yanpas) [inline elooplib/threadpool.hpp]: add a name to argument

**Review by Yanpas — CHANGES_REQUESTED** [review]

File: elooplib/eventloop.cpp (hunk 12 above)
Comment (Dekz120) [reply elooplib/eventloop.cpp]: Wee-wee
soon enouhg, sir.

File: elooplib/client.cpp (hunk 16 above)
Comment (Dekz120) [reply elooplib/client.cpp]: i need fd, to create a new Node obj(ClientTask) in eventloop

File: elooplib/threadpool.cpp (hunk 34 above)
Comment (Dekz120) [reply elooplib/threadpool.cpp:21]: quiet!

Comment (Yanpas) [reply elooplib/threadpool.cpp:21]: ehh... "execute while it's not silent"? stop/run would be better

**Comment by Yanpas** [comment]
General advice: to improve readbility add to all class member variables some prefix or suffix. E.g. instead of `nodes` use `nodes_`.

File: elooplib/client.cpp (hunk 16 above)
Comment (Yanpas) [reply elooplib/client.cpp]: Then why don't you just create it near ` Node obj(ClientTask)` in eventloop? You don't use fd here at all

File: elooplib/threadpool.cpp (hunk 34 above)
Comment (Dekz120) [reply elooplib/threadpool.cpp:21]: no. i was saing "quiet!" to you.. 
Oh shit, i'am sorry(for that). Our daddy tought us not to be ashamed about our code. Especially since they've such a good revievier and all..

File: elooplib/client.cpp (hunk 16 above)
Comment (Dekz120) [reply elooplib/client.cpp]: i need a fd to add it to pollfd. My point is `handleConnection` returns zero, or fd for new `client`( whatever it is)
I have no other idea, how to create a new client, without return code from `handleConnection`

Comment (Yanpas) [reply elooplib/client.cpp]: My argument, as well as many others was about style, not the correctness. I'll send you a voice memo with explanations

File: elooplib/archive.hpp (hunk 1 above)
Comment (Dekz120) [reply elooplib/archive.hpp:6]: is zlib a system library?

Comment (Yanpas) [reply elooplib/archive.hpp:6]: Yes, it's installed to /usr/include and may require installing extra package.

If you include with regular quotes - it means you ship zlib with your repository or it gets downloaded somewhere locally

File: Makefile [hunk 38]
```diff
@@ -1,36 +1,21 @@
...
+main.o: bvnServer/main.cpp $(INCLUDES)
+	$(CXX)  -c $(CXXFLAGS) -I$(INCLUDES) bvnServer/main.cpp
 
-eventloop.a: $(OBJFILES)
-	ar rcs eventloop.a $(OBJFILES)
+eventloop.a: $(OBJECTS) 
+	ar rcs eventloop.a $(OBJECTS)
 	
-eventloop.o: $(LSOURCES) 
+eventloop.o: $(SOURCES)
```
Comment (Yanpas) [inline Makefile]: o file can depend only on one cpp file

File: elooplib/archive.cpp [hunk 39]
```diff
@@ -0,0 +1,103 @@
+#include "archive.hpp"
+
+static constexpr size_t chunk = 16384;
+int compress(FILE *source, FILE *dest, int level)
```
Comment (Yanpas) [inline elooplib/archive.cpp:4]: I don't mind using c files, but fstreams support raii

File: elooplib/archive.hpp [hunk 40]
```diff
@@ -1,51 +1,9 @@
...
-        assert(strm.avail_in == 0);     
-
-    } while (flush != Z_FINISH);
-    assert(ret == Z_STREAM_END);       
-
-    (void)deflateEnd(&strm);
-    return Z_OK;
-}
\ No newline at end of file
+int compress(FILE *, FILE *, int);
```
Comment (Yanpas) [inline elooplib/archive.hpp:8]: Good

File: elooplib/client.cpp [hunk 41]
```diff
@@ -41,7 +37,7 @@ int Client::handleConnection()
   return res;
 }
 
-int Client::recognizeData()
+int Client::recognizeData() // Codes should be : OK or LONG_OPERATION IN TH_POOL
```
Comment (Yanpas) [inline elooplib/client.cpp]: Use enum class instead of int

File: elooplib/client.cpp [hunk 42]
```diff
@@ -6,22 +6,22 @@ Client::Client(Client &&rhs)
...
+    return nullptr;
   }
   if (rcv == 0)
   {
     Node::closeConnection();
-    return 0;
+    return nullptr;
   }
-  int res = 0;
+  std::shared_ptr<Node> res = nullptr;
```
Comment (Yanpas) [inline elooplib/client.cpp:24]: Shared ptr is null with default constructor

File: elooplib/client.cpp [hunk 43]
```diff
@@ -6,22 +6,22 @@ Client::Client(Client &&rhs)
...
 {
   char buff[512];
-  size_t rcv;
+  int rcv;
 
   rcv = recv(getFd(), buff, 512, 0);
   if (rcv < 0)
   {
-    return -1;
+    return nullptr;
```
Comment (Yanpas) [inline elooplib/client.cpp:17]: Do you know what constructor will be called here? Will it work if ctor will be `explicit`?

File: elooplib/client_task.cpp [hunk 44]
```diff
@@ -1,15 +1,22 @@
...
+  th_pool = tp;
+  Client::setFd(pc->getFd());
+  dir = pc->Client::getResponse();
+  parentClient = pc;
+  ClientTask::prepareFileTask();
+}
+
 void ClientTask::prepareFileTask()
 {
   std::function<bool(const std::string)> cond_func;
```
Comment (Yanpas) [inline elooplib/client_task.cpp:22]: I guess you forgot a reference to string

File: elooplib/eventloop.cpp [hunk 45]
```diff
@@ -1,49 +1,33 @@
 #include "eventloop.hpp"
 
-EventLoop::EventLoop(size_t n_clients) : fds(n_clients), max_pollfd_pos(0),
-                                         th_pool(new ThreadPool(3)){};
+EventLoop::EventLoop(size_t n_clients) : max_pollfd_pos(0), fds(n_clients),
+                                         nodes(n_clients, nullptr), th_pool(new ThreadPool(get_nprocs())){};
```
Comment (Yanpas) [inline elooplib/eventloop.cpp:4]: You forgot to include a header for get_nprocs

File: elooplib/tp_tasks.hpp [hunk 46]
```diff
@@ -12,7 +12,7 @@ class FileITask : public ITask
   FileITask(const std::string &, int, std::atomic<int> *, std::atomic<int> *, int);
   int handleFile(const std::string &, const std::string &,
                  std::function<int(FILE *, FILE *, int)> &);
-  virtual int doWork(){};
+  virtual int doWork(){return 0;};
```
Comment (Yanpas) [inline elooplib/tp_tasks.hpp:15]: I guess this should be pure virtual. BTW why does this class return int instead of void?

File: elooplib/eventloop.cpp (hunk 45 above)
Comment (Dekz120) [reply elooplib/eventloop.cpp:4]: i did it in hpp file, sir

Comment (Yanpas) [reply elooplib/eventloop.cpp:4]: Ah, I see. It's better to do it include it inside cpp file, because what you include in your headers is seen to everyone else. And this function - is private implantation detail

**Comment by Yanpas** [comment]
@Dekz120 , let's finish small tasks and merge

**Comment by Yanpas** [comment]
ping

**Comment by Yanpas** [comment]
Hooray!
